### PR TITLE
feat: Palette Search Ranking

### PR DESCRIPTION
**Pull request by asim-noon**
## Meta

| Change ID | Type | Confidence | Plan | Review |
|-----------|------|------------|------|--------|
| `s91a` | feat | 4.0/5.0 | 8/8 tasks, 24/24 acceptance ✓ | ✓ 1 cycle |

| Impact | +/− | Net |
|--------|----:|----:|
| raw | +1252 / −11 | +1241 |
| **true** | **+655 / −9** | **+646** |
| └ impl | +372 / −9 | +363 |
| └ tests | +283 / −0 | +283 |

<sub>excludes `fab/`, `docs/` · generated by fab-kit v2.23.11</sub>

**Pipeline:** [intake](https://github.com/sahil87/run-kit/blob/260903-s91a-palette-search-ranking/fab/changes/260903-s91a-palette-search-ranking/intake.md) ✓ → [apply](https://github.com/sahil87/run-kit/blob/260903-s91a-palette-search-ranking/fab/changes/260903-s91a-palette-search-ranking/plan.md) ✓ → review ✓ → hydrate ✓ → ship → review-pr

## Summary

The command palette filters with a plain `.includes()` over label + description and renders survivors in source-declaration order — a filter with no sort. With a registry grown to hundreds of entries, `query = "pr"` puts `Server: Protect *` and `Layout: Promote Web` above `PR: Refresh Status` and `Open: PR #3127`, and every new palette entry makes the primary discovery surface (Constitution V) worse. This ranks results by match quality instead: a pure six-tier ladder with a density → MRU → declaration-order tiebreak chain, no new dependency, and membership bit-for-bit identical to the old filter — nothing that matched before is dropped.

## Changes

- **New `lib/palette/rank.ts` (pure)** — the ranking core: Exact (whole label, or the label's `"<Category>: "` prefix) → WholeWord → WordStart → Acronym → Incidental → DescriptionOnly, with alphanumeric-run tokenization and a contiguous-initials acronym rule. Within a tier: density (`query.length / label.length`, label only) → MRU → declaration order via a stable index-decorated sort. Empty query filters nothing and orders MRU-first. Returns a new array; never mutates its input.
- **New `lib/palette/mru.ts` (storage)** — per-viewer recency under `runkit-palette-mru`, capped at 20 ids, following the `lib/last-window-per-server.ts` try/catch-noop shape so private mode or quota never throws (Constitution II/IV: no backend state, per-viewer state in localStorage).
- **`components/command-palette.tsx`** — only the plain filter arm becomes a memoized `rankActions(actions, query, mru)`; the confirm and option-picker sub-step branches are untouched (their fixed order is load-bearing for the picker's index). MRU is read once on mount and records the *originating* action id (`confirming.id` / `picking.id`), never the synthetic `-confirm` / `-opt-` row ids.
- **Tests** — new colocated `rank.test.ts` (per-tier classification, tokenizer, acronym contiguity, density, MRU, stability, non-mutation, and an end-to-end `query="pr"` regression) and `mru.test.ts` (round-trip, dedupe, cap eviction, malformed/unavailable storage); `command-palette.test.tsx` gains ranked-order, MRU-recording and membership-parity assertions.
- **Non-goals** — no change to which actions exist or how they are gated, no `PaletteAction` schema change, no new dependency, no backend persistence, `disabled` stays out of the comparator (carried as an Open Question).
- **Memory** — `docs/memory/run-kit/ui/keyboard-and-palette.md` gains § Palette search ranking plus Design Decisions entries.

## Verification

- `npx tsc --noEmit` — clean
- `just test-frontend` — 179 files / 3692 tests green
- Full `just test` — backend + frontend green. Nine e2e failures were adjudicated environmental: an isolated re-run of exactly those 6 specs passed 57/57 both with and without this change, and none of them touch the palette.

## Demo — palette-ranking

**Before** — without the feature / bug present:

https://github.com/user-attachments/assets/6fb1cdd3-1f2e-40f0-84bf-aec45e8a7553

**After** — with the feature / bug fixed:

https://github.com/user-attachments/assets/a8334108-2cdc-4ba6-8064-104032b60f6a

## Demo — palette-acronym

**Before** — without the feature / bug present:

https://github.com/user-attachments/assets/91842ebf-7167-4b79-b2ff-a008d45c70ca

**After** — with the feature / bug fixed:

https://github.com/user-attachments/assets/828dd14e-6053-41a4-b1e8-77af60781d06

### CODE REVIEW

File: app/frontend/src/components/command-palette.test.tsx [hunk 1]
```diff
@@ -520,4 +521,166 @@ describe("CommandPalette", () => {
...
+      });
+
+      expect(renderedLabels()).toEqual([
+        "PR: Refresh Status",
+        "Open: PR #3127",
+        "Server: Protect noon",
+      ]);
+    });
+
+    it("keeps membership identical to the old label+description filter", () => {
```
[the remark this answers is not in the document]
Comment (asim-noon) [reply app/frontend/src/components/command-palette.test.tsx]: Fixed — renamed the test to "keeps every action the old label+description filter admitted", so it names the superset contract (R5) rather than set equality. (ef34757)

File: app/frontend/src/lib/palette/rank.ts [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+ * the palette's filter step. Replaces "filter, then render in declaration
+ * order" with a six-tier match ladder so the strongest match — not the
+ * earliest-declared row — leads the list.
+ *
+ * Pure and dependency-free, per the lib/palette builder convention
+ * (shell.ts, zen.ts, server-protect.ts): no React, no DOM, structurally
+ * typed over `{ id, label, description? }` so it carries no component
+ * dependency. `rankActions` subsumes the old filter — membership is
+ * identical to a case-insensitive `includes` over label + description — and
+ * returns a NEW array, never mutating the caller's (upstream-memoized) list.
```
[the remark this answers is not in the document]
Comment (asim-noon) [reply app/frontend/src/lib/palette/rank.ts]: Fixed — the module docblock now states membership is a SUPERSET of the case-insensitive `includes` predicate, because the Acronym tier necessarily admits matches that predicate rejects (`ns` → `New Session`); the line-29 `Incidental` docstring was reworded in the same commit. (ef34757)

File: docs/memory/run-kit/ui/keyboard-and-palette.md [hunk 3]
```diff
@@ -340,6 +340,29 @@ The entries are folded into `paletteActions` in AppShell only. The **board route
...
+
+**The six-tier match ladder** — every matching action classifies into exactly one tier (the first tier whose rule holds wins), tiers ordered ascending (0 best); all comparisons run on lowercased strings:
+
+| Tier | Name | Rule |
+|------|------|------|
+| 0 | `Exact` | the whole label equals the query, **or** the label's `"<Category>: "` prefix equals the query — `pr` → `PR: Refresh Status` (category `PR`) |
+| 1 | `WholeWord` | some word of the label equals the query — `pr` → `Open: PR #3127` |
+| 2 | `WordStart` | some word of the label starts with the query and is longer than it — `pr` → `Server: Protect noon`, `Layout: Promote Web` |
+| 3 | `Acronym` | the query is a contiguous run of the label's word-initials string — `prs` → `PR: Refresh Status` (`ps` does NOT match: the initials run must be contiguous) |
+| 4 | `Incidental` | the query occurs in the label, but only strictly inside a word |
```
[the remark this answers is not in the document]
Comment (asim-noon) [reply docs/memory/run-kit/ui/keyboard-and-palette.md]: Fixed — the tier-4 row now describes the actual rule (label contains the query and no higher tier applied), notes that an alphanumeric query is then necessarily strictly inside a word, and gives the boundary-spanning case `kit: r` → `run-kit: Restart Daemon`. (ef34757)

File: fab/changes/260903-s91a-palette-search-ranking/intake.md [hunk 4]
```diff
@@ -0,0 +1,232 @@
...
+
+**Tier definitions** (all comparisons on `toLowerCase()`ed strings; the raw label is never mutated):
+
+| Tier | Rule | `query = "pr"` examples |
+|------|------|-------------------------|
+| 0 Exact | `label === query`, **or** the label has a `"<Category>: "` prefix and `category === query` | `PR: Refresh Status` (category `PR`) |
+| 1 WholeWord | some word of the label equals the query exactly | `Open: PR #3127` (word `PR`) |
+| 2 WordStart | some word of the label starts with the query but is longer than it | `Server: Protect noon`, `Layout: Promote Web` |
+| 3 Acronym | the query's letters equal the first letters of a **consecutive run** of words | `prs` → `PR: Refresh Status`; `pw` → `Layout: Promote Web` |
+| 4 Incidental | the query occurs in the label, but only strictly inside a word | `pr` → `Web: Reprint` (hypothetical) |
```
[the remark this answers is not in the document]
Comment (asim-noon) [reply fab/changes/260903-s91a-palette-search-ranking/intake.md]: Fixed — the Intake tier table row 4 now matches the implemented catch-all rule and carries the boundary-spanning example. (ef34757)

File: fab/changes/260903-s91a-palette-search-ranking/plan.md [hunk 5]
```diff
@@ -0,0 +1,257 @@
...
+#### R1: Six-tier match ladder
+`rankActions` SHALL classify every action that matches the query into exactly one tier, and SHALL order tiers ascending (0 best). The tiers, evaluated against the lowercased label unless stated otherwise, are:
+
+| Tier | Name | Rule |
+|------|------|------|
+| 0 | Exact | the whole label equals the query, **or** the label carries a `"<Category>: "` prefix whose category equals the query |
+| 1 | WholeWord | some word of the label equals the query |
+| 2 | WordStart | some word of the label starts with the query and is longer than it |
+| 3 | Acronym | the query is a contiguous substring of the label's word-initials string |
+| 4 | Incidental | the query occurs in the label, but only strictly inside a word |
```
[the remark this answers is not in the document]
Comment (asim-noon) [reply fab/changes/260903-s91a-palette-search-ranking/plan.md]: Fixed — R1 tier 4 now states the implemented rule and calls out that the catch-all is exactly what keeps R5's membership superset intact. The implementation is correct here and the spec prose was the defective side: tightening `Incidental` to a strict inside-a-word test would drop boundary-spanning queries (`kit: r`, `refresh st`) that the pre-change `includes` filter admitted, violating R5. Also added a `matchTier` unit test locking the boundary-spanning case. (ef34757)
